tests/validation.py

```diff
@@ -0,0 +1,166 @@
+from unittest import TestCase
+
+from adapters import DeclarativeValidationNode, ValidationError, ValidationTree
+
+
+class ValidatorsTest(TestCase):
+    def assertHasValidationError(self, result, message, field=None):
+        self.assertFalse(result.is_valid)
+        error = ValidationError(message, field)
+        self.assertIn(error, result.errors)
+
+    def _test_username_tree(self, tree):
+        self.assertHasValidationError(
+            tree.validate({}),
+            'missing data: username'
+        )
+        self.assertHasValidationError(
+            tree.validate({'username': 'foobar'}),
+            'invalid username', 'username'
```

Comment:
Keyword arguments would make this clearer I think:

```python
self.assertHasValidationError)
    tree.validate({'username': 'foobar'},
    message='invalid username',
    field='username',
)
```

Reply:
For me it makes sense to expect the single word string to be field name and phrase string to be message.
But I can live with both implementations :)